src/simple_types.rs

```diff
@@ -1,34 +1,34 @@
 use std::iter::Sum;
 
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub(crate) struct OrderLineId {
-    value: String,
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub(crate) struct OrderLineId<'a> {
+    value: &'a str,
 }
 
-impl OrderLineId {
-    pub(crate) fn new(id: impl Into<String>) -> Self {
-        Self { value: id.into() }
+impl<'a> OrderLineId<'a> {
+    pub(crate) fn new(id: &'a str) -> Self {
+        Self { value: id }
     }
 }
-#[derive(Clone)]
-pub(crate) struct OrderId {
-    value: String,
+#[derive(Clone, Copy)]
+pub(crate) struct OrderId<'a> {
```

Comment:
I started using something like https://github.com/zbraniecki/tinystr 
and ensure it is struct allocated, so copy is just as fast :) 

Lifetime explosion can be a pain to deal with. 

Long time no see 😛

Reply:
There is also https://lib.rs/crates/shared-string perhaps more suited for larger strings.

Reply:
Thank you for the help! These crates seem really useful, might see if they're fit for my work stuff too. 

What's brought you to this repo?

Reply:
Was checking out Rust to get a feel for how productive I can be. I ended up doing a port of my pet Scala project to Rust for perf reasons 😄 

Was looking for ideas and design patterns. 😄

Reply:
You could use macros for some of the type constraints too. I have created a macro which auto generates, `Mul`, `Div`, `From`, `To` impls

```rust
num_new_type_rules![
  // commutative
  UnitPrice * Quantity => Amount
  Quantity * Amount => Amount

  // division
  Amount / UnitPrice => Quantity
  Amount / Quantity => UnitPrice

  // from-into conversion
  Amount <=> Quantity
  Amount <=> UnitPrice
];
```